lib/task.js

```diff
@@ -1,155 +1,123 @@
-      dep(cb);
-    } else {
-      cb();
-    }
-  }
+  run(options) {
+    let finished = false;
+    const orig = Object.assign({}, this.options);
+    this.options = Object.assign({}, this.options, options);
+    this.status = 'preparing';
+    this.emit('preparing', this);
 
-  /**
-   * Run a task, capture it's start and end times
-   * and emitting 3 possible events:
-   *
-   *  - starting: just before the task starts with `task` and `run` objects passed
-   *  - finished: just after the task finishes with `task` and `run` objects passed
-   *  - error: when an error occurs with `err`, `task`, and `run` objects passed.
-   *
-   * @param  {Function} `cb` Callback function to be called when task finishes running.
-   * @return {*} undefined, `Promise` or `Stream` to be used to determine when task finishes running.
-   */
-
-  run(next) {
-    // exit early when task set not to run.
-    if (skip(this)) {
-      next();
-      return;
+    if (this.skip(options)) {
+      return () => Promise.resolve(null);
     }
 
-    const cb = this.wrapCallback(next);
-    this.runDeps(err => {
-      if (err) {
-        next(err);
-        return;
-      }
-
-      try {
-        const callback = this.callback;
-        const results = utils.isGenerator(callback)
-          ? co(callback, cb)
-          : callback.call(this, cb);
-
-        // needed to capture when the actual task is finished running
-        if (typeOf(results) === 'object' && typeof results.on === 'function') {
-          results.on('error', cb);
-          results.on('end', cb);
-          results.resume();
+    this.time = new Timer();
+    this.time.start();
+    this.status = 'starting';
+    this.emit('starting', this);
+
+    return () => {
+      return new Promise((resolve, reject) => {
+        const finish = (err, val) => {
+          if (finished) return;
+          finished = true;
+          this.options = orig;
+          this.time.end();
+          this.status = 'finished';
+          this.emit('finished', this);
+          if (err) {
+            utils.define(err, 'task', this);
+            this.emit('error', err);
+            reject(err);
+          } else {
+            resolve(val);
+          }
+        };
+
+        if (typeof this.callback !== 'function') {
+          finish();
           return;
         }
 
-        if (typeOf(results) === 'promise') {
-          results.then(res => cb(null, res)).catch(cb);
+        try {
+          const res = this.callback.call(this, (err, val) => {
+            if (err) {
+              finish(err);
+            } else {
+              finish(null, val);
+            }
+          });
+
+          if (isEmitter(res)) {
+            res.once('error', finish);
+            res.once('finish', finish);
+            res.once('end', finish);
+            return;
+          }
+
+          if (this.callback.length === 0) {
+            if (res.then) {
+              res.then(() => finish());
```

Comment:
This line could be `res.then(val => finish(null, val));` so if any values are returned from the promise they'll be passed through like with callbacks.